openupgrade_scripts/scripts/sale/16.0.1.2/pre-migration.py

```diff
@@ -0,0 +1,147 @@
+    which has analytic.tag then form it as a jsonb like {'1': 100, '2': 50}
+    and also check if the table has analytic_account_column then check if it
+    exist in the analytic_account of all the analytic_distribution of the analytic tags
+    then take it as the 100%, which mean an account.move.line both specify '2' as the
+    analytic.account.id and it has 1 analytic.tag also have that analytic.account then
+    the value will sum together
+    """
+    if not openupgrade.column_exists(
+        env.cr, "sale_order_line", "analytic_distribution"
+    ):
+        openupgrade.logged_query(
+            env.cr,
+            """
+            ALTER TABLE sale_order_line
+            ADD COLUMN IF NOT EXISTS analytic_distribution jsonb;
+            """,
+        )
+
+    openupgrade.logged_query(
+        env.cr,
+        """
+        WITH distribution_data AS (
+            WITH sub AS (
+                SELECT
+                    all_line_data.sale_order_line_id,
+                    all_line_data.analytic_account_id,
+                    SUM(all_line_data.percentage) AS percentage
+                FROM (
+                    SELECT
+                        so_line.id AS sale_order_line_id,
+                        account.id AS analytic_account_id,
+                        100 AS percentage
+                    FROM sale_order_line so_line
+                    JOIN sale_order so ON so.id = so_line.order_id
+                    JOIN account_analytic_account account
+                        ON account.id = so.analytic_account_id
+                    WHERE so.analytic_account_id IS NOT NULL
+
+                    UNION ALL
+
+                    SELECT
+                        so_line.id AS sale_order_line_id,
+                        dist.account_id AS analytic_account_id,
+                        dist.percentage AS percentage
+                    FROM sale_order_line so_line
+                    JOIN account_analytic_tag_sale_order_line_rel tag_rel
+                        ON tag_rel.sale_order_line_id = so_line.id
+                    JOIN account_analytic_distribution dist
+                        ON dist.tag_id = tag_rel.account_analytic_tag_id
+                    JOIN account_analytic_tag aat
+                        ON aat.id = tag_rel.account_analytic_tag_id
+                    WHERE aat.active_analytic_distribution = true
+                ) AS all_line_data
+                GROUP BY all_line_data.sale_order_line_id, all_line_data.analytic_account_id
+            )
+            SELECT
+                sub.sale_order_line_id,
+                jsonb_object_agg(sub.analytic_account_id::text, sub.percentage)
+                    AS analytic_distribution
+            FROM sub
+            GROUP BY sub.sale_order_line_id
+        )
+
+        UPDATE sale_order_line so_line SET analytic_distribution = dist.analytic_distribution
+        FROM distribution_data dist WHERE so_line.id = dist.sale_order_line_id
+        """,
```

Comment:
This might be problematic to fill for most cases. I tested it with a sale order which had a `analytic_account_id` set before the migration. After the migration I see the `analytic_distribution` correctly (100% for the analytic account). and `analytic_account_id` is still set.

When I create an invoice from it this results in 200% for the analytic account because both are set in the sale.order. In `_prepare_invoice_line` Odoo actually adds 100%. This is the reason why it would work without a migration. Alternativly you might want to clear the `analytic_account_id` on the sale.order to prevent the +100%.

```sql
UPDATE sale_order SET analytic_account_id = NULL
WHERE analytic_account_id IS NOT NULL
```

Reply:
I think then it's better to not do anything at all and leave it as it was.

Reply:
I rechecked the query. It might also be worth to remove the lines 80-90 (first part of UNION ALL) of the query because the 100% are added there to the distribution. Fully skipping it would mean that you also skip the 2nd part of the query and don't work against the removal of `analytic_tag_ids` on `sale.order.line`.

Reply:
@fkantelberg Can you provide an example that this query doesn't work please, a video would help if you have one.

Reply:
I don't know about video. Here is the an example how it looks in the database. Following is 15.0:
```
>>> so.read(["name", "state", "analytic_account_id"])
[{'id': 1,
  'name': 'AG-2024-01-0001',
  'state': 'sale',
  'analytic_account_id': (1, 'Intern')}]

>>> so.order_line.read(["order_id", "product_id", "analytic_tag_ids"])
[{'id': 1,
  'order_id': (1, 'AG-2024-01-0001'),
  'product_id': (4, 'Test'),
  'analytic_tag_ids': []}]
```

After the migration it looks like this:
```
>>> so.read(["name", "analytic_account_id"])
[{'id': 1, 'name': 'AG-2024-01-0001', 'analytic_account_id': (1, 'Intern')}]
>>> so.order_line.read(["order_id", "product_id", "analytic_distribution"])
[{'id': 1,
  'order_id': (1, 'AG-2024-01-0001'),
  'product_id': (4, 'Test'),
  'analytic_distribution': {'1': 100}}]
```

If I confirm the order and create an invoice out of it I see that the invoice has a wrong analytic distribution because the `analytic_account_id` on the `sale.order` is adding +100.
```
>>> so.mapped("invoice_ids.invoice_line_ids").read(["move_id", "analytic_distribution"])
[{'id': 1,
  'move_id': (1, 'Draft RE-2023-12-0001'),
  'analytic_distribution': {'1': 200.0}}]
```

Reply:
@fkantelberg ok you seem right although i think odoo should remove the `analytic_account_id` too just like what they did for purchase. I have remove the code from line 80 - 90 as you suggested. Please retest if needed
cc @pedrobaeza

Reply:
Works flawlessly and results in a correct invoice.

Regarding `analytic_account_id` there is still code in `action_confirm` which creates an analytic account for expense products for a sale.order.